### DISCUSSION TITLE
How to restrict access to a publication's resources (e.g. a sample)

### DISCUSSION THREAD

**Discussion by RahimMahmoudzadeh** [readium/kotlin-toolkit#99] [Q&A]
Greetings 
I want to show my user only 10 pages of the book, for example, I wanted you to help me in this case, how can I show a preview of the book to the user?

**Comment by mrifni** [comment]
You better off creating a sample epub and a separate full epub file.

**Reply by RahimMahmoudzadeh** [reply]
Can you explain more about that?
Can not be with Readium?

**Comment by mickael-menu** [comment]
Like @mrifni mentioned, the most reliable and safest way would be to create your own sample EPUBs. No technique in Readium can fully prevent a user from accessing the full content.

However, you have several options to do this in Readium:

* (easy) If you just want to prevent accessing some resources in the EPUB, set a custom `openPublication` callback when using `Streamer.open()` and remove the links you want to restrict the access to from `manifest.readingOrder`.
* (easy) Observe the `navigator.currentLocator` value and when you receive a `Locator` that exceeds the allowed portion (e.g. checking `locations.position` or `locations.totalProgression`, then hide the navigator and show a message explaining that the user reached the end of the sample.
* (hard but safer) If you need to restrict a portion of a resource, implement a custom `Fetcher` and modify the HTML page on the fly when it is requested.

**Reply by RahimMahmoudzadeh** [reply]
Greetings
Thank you for your answer
I now want to use method 3

But I wanted to know how to implement a custom Fetcher and get the text in html and break them
Because now I can tell Readium to show, for example, only 3 of the chapters
But some chapters have a lot of text, I want to break that.

And did I understand correctly that the html operation is performed in the HtmlInjector class?

And is there a better way to delete some of the texts in the chapter at all?

Thanks again for your guidance

**Reply by mickael-menu** [reply]
No the `HtmlInjector` is a private implementation details, so you should not modify it.

You can wrap the default fetcher with your own `TransformingFetcher`, passing it a function that will modify the `Resource` on the fly:

```kotlin
val publication = streamer.open(
    asset,
    onCreatePublication = {
        fetcher = TransformingFetcher(fetcher, ::stripHtml)
    }
).getOrThrow()

fun stripHtml(resource: Resource): Resource =
    object : TransformingResource(resource) {
        override suspend fun transform(data: ResourceTry<ByteArray>): ResourceTry<ByteArray> =
            if (!resource.link().mediaType.isHtml)
                data
            else
                data.mapCatching {
                    val html = it.toString()
                    // Modify the HTML here...
                    html.toByteArray()
                }
    }
```
